src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
+// a side effect. On success, a new MulNode is returned.
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, BasicType bt) {
+  // We need to make sure that the current AddNode is not part of a MulNode that has already been optimized to a
+  // power-of-2 addition (e.g., 3 * a => (a << 2) + a). Without this check, GVN would keep trying to optimize the same
+  // node and can't progress. For example, 3 * a => (a << 2) + a => 3 * a => (a << 2) + a => ...
+  if (find_power_of_two_addition_pattern(this, bt, nullptr) != nullptr) {
```

Comment:
Where does the optimization `3 * a => (a << 2) + a` happen? Do we use `find_power_of_two_addition_pattern` there too? If not: how do we prevent the code of the two locations from diverging in the future?

Reply:
This `3 * a => (a << 2) + a` happens in `MulINode::Ideal()` and is independent from my code. My code does not explicitly produce power of two patterns but simple multiplication nodes, but such a multiplication node will be later idealized into power of two's. 

Power of two patterns make my code more complex, but this is out of my control.

> how do we prevent the code of the two locations from diverging in the future?

I don't expect primitive opts like power-of-two being removed. Even if it does, `find_simple_multiplication_pattern()` will still work.